### ISSUE TITLE
Question about CloudKitProcedure error handling

### ISSUE THREAD

**Issue by leotumwattana** [ProcedureKit/ProcedureKit#618]
I understand that `CloudKitProcedure` is a subclass of `RetryProcedure` which I guess makes retrying for certain types of *CloudKit errors* a bit easier. However, for the case of `.limitExceeded`, it seems the api is a bit convoluted to me.

Since handling `.limitExceeded` requires not just "one" re-configured procedure (not counting the delay procedure), but at least two (e.g. bisected), then returning the suggested configuration block seems a bit strange. If the framework consumer wants to do some even more complex custom handling, then this suggested configuration block seems a bit out of place?

I personally am tempted to just `return nil` in the custom error handler and just use `addChildren`. Haven't tried it yet and don't know if that'll work and whether there are unintended consequences such as the wrong errors getting propagated through.

Which leaves me thinking, why not allow the framework user to return an array of procedures to retry? Maybe the suggested configuration block can simply be used for applying to the procedure that is to be retried? The suggested block will just return the necessary procedures configured as suggested, and the framework consumer can tag on any other procedures as necessary?

Anyway, just my two cent.

**Comment by danthorpe** [comment]
Hi @leotumwattana  - some really good points here, both in terms of `RetryProcedure` and `CloudKitProcedure`. I've been working on #582 which adds automatic "bisecting" and retrying when `.limitExceeded` error is received. But, this has highlighted a deficiency in how `RetryProcedure` operates which you're also alluding to here.

Basically, `RetryProcedure` works like this:

1. Execute the first operation out of the iterator

2. Use `GroupProcedure`'s error handling to intercept operations of type `T` which finished with errors, and add (i.e. retry) the next value from the iterator.

3. `GroupProcedure` will mark a failed operation as "recovered" if another instance of `T` succeeds.

So, `RetryProcedure` essentially is just a `GroupProcedure` with an iterator to feed new values of operations. 

Now, in pretty much most cases, this is fine. e.g. Networking operations - just retry the same procedure.

However, for `CloudKitProcedure<T>`, and `.limitExceeded` error handling, we basically want to do something like this: 

1. Take whatever should be "processed", and bisect it into 2 halves - left and right. 

2. Create a new procedure to process the left

3. Configure the retried procedure to do the right.

4. The right procedure now also depends on the left procedure.

The problem with this, is that when the LHS procedure finishes without any errors - it stops processing the right hand side.

So, I think, maybe a re-think of how `RetryProcedure` and maybe even `GroupProcedure` works is in order...

**Comment by danthorpe** [comment]
Part of all this is so that `GroupProcedure` subclasses have a mechanism which allows them to purge collected errors from children. At the moment, this mechanism is quite weak, and not at all explicit.

So, possibly, `GroupProcedure` needs APIs which specifically allow for indicating which operations have now been recovered. E.g. Operation A finished with and error - i.e. the group has failures. But, the subclass will add operations B, C & D to the group, and if D finishes with an error - we can remove the failure from A.

**Comment by leotumwattana** [comment]
Yup. Just what I'm thinking, but put much more eloquently. 

Maybe the APIs for removing failures should be part of `RetryProcedure` instead?

**Comment by danthorpe** [comment]
Yeah, maybe. I think for the moment, (Beta 5) this will not change. But I'll look at re-writing parts of `GroupProcedure`, and `RetryProcedure` next week.

**Comment by leotumwattana** [comment]
Sounds good. Thanks @danthrope! 👍

**Comment by briancordanyoung** [comment]
I am working to improve the error handling in a complex ProcedureKit based app.  At this point I'm really unclear on the best way to do this.

Most user actions cause the creation of a custom GroupProcedure.  Not all child operations are required to complete without errors.  In these cases, an error just means an optional item doesn't exist. Is there currently a way to intercept the error from the child operation, and if acceptable, block it from being added to the GroupProcedure error array?  I don't want the Parent Group operation to report these specific errors.

The way I read the thread above, this is what @danthorpe & @leotumwattana are proposing to add to GroupProcedures?

**Comment by swiftlyfalling** [comment]
@briancordanyoung: There is currently some functionality in `GroupProcedure` to support your use-case. Specifically, you can add the following override in your GroupProcedure subclass:

```swift
override open func child(_ child: Operation, willAttemptRecoveryFromErrors errors: [Error]) -> Bool {
    // based on the child operation that is finishing with errors, determine whether to "recover" from those errors (i.e. exclude them from "fatal" errors)
    if child === ignoredProcedure {
        // if necessary, check its error (to verify that the reported error is one that can be ignored)
        // if the error can be ignored, then return true
        return true // this child's errors will not be treated as "fatal", and will not be included in the Group's errors
    }
}
```

This functionality currently supports `RetryProcedure` and should, I think, do the trick for your use-case.

**Comment by danthorpe** [comment]
Yup - @swiftlyfalling is exactly right.

The APIs were written from the perspective of error recovery or failure. Recovery means _add another procedure instance which might not fail_ to the group. But you could also just absorb the error and not add another procedure, in essence ignoring it.

Always open to suggestion on how this could be further improved though.
